pkgs/cli_util/lib/src/base_directories.dart

```diff
@@ -0,0 +1,274 @@
+
+import '../cli_util.dart';
+
+/// The standard system paths for a Dart [tool].
+///
+/// This class respects the following directory standards:
+///
+/// - On Linux, the [XDG Base Directory
+///   Specification](https://specifications.freedesktop.org/basedir-spec/latest/).
+/// - On MacOS, the
+///   [Library](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html)
+///   directory.
+/// - On Windows, `%APPDATA%` and `%LOCALAPPDATA%`.
+///
+/// Note that [cacheHome], [configHome], [dataHome], [runtimeHome], and
+/// [stateHome] may be overlapping or nested.
+///
+/// Note that the directories won't be created, the methods merely return the
+/// recommended locations.
+final class BaseDirectories {
+  /// The name of the Dart tool.
+  ///
+  /// This should be a valid directory name on every operating system,
+  /// typically containing only lower-case characters and underscores to ensure
+  /// maximum compatibility and readability. For example, "my_dart_tool".
+  final String tool;
+
+  /// The environment variables to use.
+  ///
+  /// Defaults to [Platform.environment].
+  final Map<String, String> environment;
+
+  /// Constructs a [BaseDirectories] instance for the given [tool] name.
+  BaseDirectories({
+    required this.tool,
+    Map<String, String>? environment,
+  }) : environment = environment ?? Platform.environment;
+
+  /// The path of the directory where [tool] can place its caches.
+  ///
+  /// The cache might at any point be purged by the operating system or user.
+  /// Applications should be able to reconstruct any data stored here. If [tool]
+  /// cannot handle data being purged, use [runtimeHome] or [dataHome] instead.
+  ///
+  /// This is a location appropriate for storing non-essential files that may be
+  /// removed at any point.
+  ///
+  /// The directory location depends on the current [Platform.operatingSystem]:
+  /// - on **Windows**:
+  ///   - `%LOCALAPPDATA%\<tool>`
+  /// - on **Mac OS**:
+  ///   - `$HOME/Library/Caches/<tool>`
+  /// - on **Linux**:
+  ///   - `$XDG_CACHE_HOME/<tool>` if `$XDG_CACHE_HOME` is defined, and
+  ///   - `$HOME/.cache/<tool>` otherwise.
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap with [runtimeHome] and
+  /// [stateHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String cacheHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.cache), tool);
+
+  /// The path of the directory where [tool] can place its configuration.
+  ///
+  /// The configuration may be synchronized across devices by the OS and may
+  /// survive application removal.
+  ///
+  /// This is a location appropriate for storing application specific
+  /// configuration for the current user.
```

Comment:
Document `plistFiles` by itself.

```dart
  /// If [plistFiles] is set to `true` on a MacOS platform, the MacOS-specific
  /// configuration directory, which must only contain `.plist` files,
  /// is returned, otherwise a directory that allows non-`.plist` files is returned.
  /// It's the tool's responsibility to only write `.plist` files when those are required.
```
-ish.